private/buf/bufgen/generator.go

```diff
@@ -173,239 +157,148 @@ func (g *generator) generateConcurrently(
+					ctx,
+					container,
+					image,
+					currentPluginConfig,
+					// TODO: We need to support the includeImports option here.
+				)
+				if err != nil {
+					return err
+				}
+				responses[index] = response
+				return nil
+			})
 		}
-	}
-	// Limit concurrent execution of remote plugins using separate semaphore channel.
-	// This time we can use a higher concurrency limit since the work is almost
-	// exclusively I/O bound.
-	remoteSemaphore := newSemaphore(thread.Parallelism() * 10)
-	defer remoteSemaphore.Discard()
-	for remote, plugins := range remoteToPlugins {
-		// prevent issues with asynchronously executed closures
-		remote := remote
-		plugins := plugins
-		// Add a job for each remote.
-		eg.Go(func() error {
-			if err := remoteSemaphore.Acquire(ctx); err != nil {
-				return fmt.Errorf("failed to acquire semaphore: %w", err)
-			}
-			defer remoteSemaphore.Release(ctx)
-			generateService, err := g.registryProvider.NewGenerateService(ctx, remote)
-			if err != nil {
-				return fmt.Errorf("failed to create generate service for remote %q: %w", remote, err)
-			}
-			references := make([]*registryv1alpha1.PluginReference, 0, len(plugins))
-			for _, pluginOption := range plugins {
-				references = append(references, pluginOption.reference)
-			}
-			responses, _, err := generateService.GeneratePlugins(ctx, bufimage.ImageToProtoImage(image), references)
-			if err != nil {
-				return err
-			}
-			if len(responses) != len(references) {
-				return fmt.Errorf("unexpected number of responses received, got %d, wanted %d", len(responses), len(references))
-			}
-			for i, response := range responses {
-				// Map each plugin response back to the right place.
-				// Note: does not require a lock, since each response is
-				// assigned to its own index in the slice.
-				pluginResponses[plugins[i].index] = bufplugin.NewPluginResponse(
-					response,
-					plugins[i].pluginRemote,
+		if pluginConfig.Name != "" {
+			jobs = append(jobs, func(ctx context.Context) error {
+				response, err := g.execLocalPlugin(
+					ctx,
+					container,
+					g.appprotoosGenerator,
+					image,
+					imagesByDir,
+					currentPluginConfig,
+					includeImports,
 				)
-			}
-			return nil
-		})
+				if err != nil {
+					return err
+				}
+				responses[index] = response
+				return nil
+			})
+		}
 	}
-	if err := eg.Wait(); err != nil {
+	// We execute all of the jobs in parallel, but apply them in order so that any
+	// insertion points are handled correctly.
+	//
+	// For example,
+	//
+	//  # buf.gen.yaml
+	//  version: v1
+	//  plugins:
+	//    - remote: buf.build/org/plugins/insertion-point-receiver
+	//      out: gen/proto
+	//    - name: insertion-point-writer
+	//      out: gen/proto
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	if err := thread.Parallelize(
+		ctx,
+		jobs,
+		thread.ParallelizeWithCancel(cancel),
+		thread.ParallelizeWithFastFail(), // We want error messages to be concise here.
+	); err != nil {
```

Comment:
These are interesting points, but I'd argue that the tradeoff of reinventing our own parallelization mechanism for this particular place in our codebase isn't worth it. One of our biggest priorities is consistency - `thread.Parallelize` might be susceptible to some of the issues you're describing, but we shouldn't go against the grain until we have a compelling reason to do so. By leaning into the `thread` package, we centralize all of our concurrency requirements in single place, where it's far more dependable and predictable.

In short, I'd prefer to remain as consistent, simple, and readable as possible until we hit a point where inventing something different is actually necessary. I could very well be wrong and that we'll need to refactor this later, but I think it's important that we optimize for consistency and don't prematurely optimize something that might not be an issue in practice.

Reply:
I think that's reasonable, but I worry that we have little observability into knowing whether this becomes a problem. Do we wait until users complain that remote plugins are undesireable because the take a long time? My biggest concern is making sure that the remote plugin requests are sent first, if we can do that and keep using `thread.Parallelize` it would be great. Maybe we should set a higher concurrency threshold for this execution?

Reply:
We're optimizing for correct code over optimized code here - the primary purpose of this PR is to fix some outstanding bugs. We've QA'ed this, if we have an issue we'll revisit.